File: fastapi-poc/barc/barber/domain/time_range.py
from typing import Iterator
from dataclasses import dataclass
from barc.barber.domain.time_of_day import TimeOfDay


@dataclass
class TimeRange:
    start: TimeOfDay
    end: TimeOfDay

    @classmethod
    def from_string(cls, x: str) -> "TimeRange":
        start, end = x.split("-")
        return cls(TimeOfDay.from_string(start), TimeOfDay.from_string(end))

    def spread(self, step=30) -> Iterator[TimeOfDay]:
        current = self.start
        times = []
        while current < self.end:
            times.append(current)
            current = current.add_minutes(step)
            if len(times) > 24 * 60 / step:
                return times
        return times

    def is_between(self, time: TimeOfDay) -> bool:
        return self.start < time < self.end

    def __str__(self) -> str:
        return f"{self.start}-{self.end}"

File: next-poc/src/time-range.ts
import {
  Time,
  addMinutes,
  buildTimeFromString,
  isBefore,
  timeToString,
} from "./time";

export type TimeRange = {
  start: Time;
  end: Time;
};

export function spread(timeRange: TimeRange, step = 30): Time[] {
  const hours = [];
  let current = timeRange.start;
  while (isBefore(current, timeRange.end)) {
    hours.push(current);
    current = addMinutes(current, step);
    if (hours.length > (24 * 60) / step) {
      return hours;
    }
  }
  return hours;
}

export function newTimeRange(start: Time, end: Time) {
  return { start, end };
}

export function fromString(str: string): TimeRange {
  const [start, end] = str.split("-").map(buildTimeFromString);
  return newTimeRange(start, end);
}

export function timeRangeToString(timeRange: TimeRange): string {
  return `${timeToString(timeRange.start)}-${timeToString(timeRange.end)}`;
}
